internal/engine/procedures/typing/typing.go

```diff
@@ -0,0 +1,663 @@
+// package typing checks the type safety of a procedure.
+// It will return data types for expressions, map[string]DataType for loop terms,
+// and nothing for statements.
+package typing
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/kwilteam/kwil-db/core/types"
+	"github.com/kwilteam/kwil-db/internal/engine"
+	"github.com/kwilteam/kwil-db/internal/engine/sqlanalyzer/typing"
+	parser "github.com/kwilteam/kwil-db/internal/parse/procedure"
+	"github.com/kwilteam/kwil-db/internal/parse/sql/tree"
+)
+
+func EnsureTyping(stmts []parser.Statement, procedure *types.Procedure, schema *types.Schema, cleanedInputs []*types.NamedType,
+	sessionVars map[string]*types.DataType) (anonReceiverTypes []*types.DataType, err error) {
+
+	declarations := make(map[string]*types.DataType)
+	for _, param := range cleanedInputs {
+		declarations[param.Name] = param.Type
+	}
+
+	for v, typ := range sessionVars {
+		_, ok := declarations[v]
+		if ok {
+			// this should never happen, since session variables have a unique
+			// prefix
+			panic(fmt.Sprintf("session variable %s collision", v))
```

Comment:
And the user is prohibited by the antlr rules from declaring a variable starting with `_`?
Can we just return an error since there's no `recover` to catch this?